library/alloc/src/collections/vec_deque/spec_extend.rs

```diff
@@ -0,0 +1,73 @@
+use crate::alloc::Allocator;
+use crate::vec::IntoIter;
+use core::slice;
+
+use super::VecDeque;
+
+// Specialization trait used for VecDeque::extend
+pub(super) trait SpecExtend<T, I> {
+    fn spec_extend(&mut self, iter: I);
+}
+
+impl<T, I, A: Allocator> SpecExtend<T, I> for VecDeque<T, A>
+where
+    I: Iterator<Item = T>,
+{
+    default fn spec_extend(&mut self, mut iter: I) {
+        // This function should be the moral equivalent of:
+        //
+        //      for item in iter {
+        //          self.push_back(item);
+        //      }
+        while let Some(element) = iter.next() {
+            if self.len() == self.capacity() {
+                let (lower, _) = iter.size_hint();
+                self.reserve(lower.saturating_add(1));
+            }
+
+            let head = self.head;
+            self.head = self.wrap_add(self.head, 1);
+            unsafe {
+                self.buffer_write(head, element);
+            }
+        }
+    }
+}
+
+impl<T, A: Allocator> SpecExtend<T, IntoIter<T>> for VecDeque<T, A> {
```

Comment:
This `IntoIter` refers to `vec::IntoIter`. Since there also is `vec_deque::IntoIter` and we're in the context of `VecDeque` I think it's better to only import the module and not the struct itself. I.e. refer to it as `vec::IntoIter` to make it easier for humans who don't look at imports.

Reply:
Done